wrote methods for 17 metrics out of a presented list and added a prototype xml converter

filipp05 opened the pull request:
**1) Made up the following metrics' calculation methods using ast python lib:**

— Number of Classes
— Number of Methods
— Number of Static Methods
— Maximum Number of Method Parameters
— Maximum Method Length
— Number of Decorators
— Number of Public Constants in File
— Number of Libraries
— Number of Extensions in the Project
— Coupling Between Objects
— Number of Files in the Project
— Depth of the Project File System Tree
— Average Number of Lines per File
— Average Number of Lines per Method
— Average Number of Methods per Class
— Average Number of Parameters per Method or Function
— Number of deprecated methods [requires more attention to be done correctly]
— Number of Functions or Methods Without Docstrings
— Number of Functions or Methods Without Typing
— Number of Context Managers
— Number of Handled Exceptions

**2) What's more, with this commit, after you run main.py, you get an xml file with the calculated metrics:
for instance, this is my local Django repo that I mostly used for primary tests**

`<?xml version="1.0" ?>
<report>
 <report-time>02.02.2025</report-time>
 <repository-path>C:\Users\ASUS\PycharmProjects\Django\blogengine</repository-path>
 <metrics>
  <metric name="Number of Classes">23</metric>
  <metric name="Number of Methods">19</metric>
  <metric name="Number of Static Methods">0</metric>
  <metric name="Maximum Number of Method Parameters">3</metric>
  <metric name="Maximum Method Length">28</metric>
  <metric name="Number of Decorators">0</metric>
  <metric name="Number of Public Constants in File">44</metric>
  <metric name="Number of Libraries">26</metric>
  <metric name="Number of Extensions in the Project">{'.sqlite3', '.css', '.py', '.xml', '.pyc', '.html', '.iml'}</metric>
  <metric name="Coupling Between Objects">{'BlogConfig': 1, 'TagForm': 6, 'PostForm': 6, 'Meta': 0, 'Post': 10, 'Tag': 4, 'ObjectDetailMixin': 3, 'ObjectCreateMixin': 5, 'ObjectUpdateMixin': 6, 'ObjectDeleteMixin': 5, 'PostCreate': 3, 'PostDelete': 3, 'TagDelete': 3, 'TagCreate': 2, 'TagUpdate': 3, 'PostUpdate': 3, 'PostDetail': 2, 'TagDetail': 2, 'Migration': 7}</metric>
  <metric name="Number of Files in the Project">73</metric>
  <metric name="Depth of the Project File System Tree">5</metric>
  <metric name="Average Number of Lines per File">29.57894736842105</metric>
  <metric name="Average Number of Lines per Method">4.75</metric>
  <metric name="Average Number of Methods per Class">0.8260869565217391</metric>
  <metric name="Average Number of Parameters per Method or Function">1.4583333333333333</metric>
  <metric name="Number of deprecated methods">[]</metric>
 </metrics>
</report>`

filipp05 commented:
As for the current commit, I added a few more metrics such as:
— Number of Functions or Methods Without Docstrings
— Number of Functions or Methods Without Typing
— Number of Context Managers
— Number of Handled Exceptions

**What is more, I had recently found a bug: we should not walk through python files in virtual environment so I have put some popular names for virtual env in config.py file. However, this may require more effort to be more flexible**

Last but not least, I used a simple exception handler that will become more complexed in advance

I am also thing of using my Metrics class as a return attribute for each metric-method

mximp commented on main.py:
```diff
@@ -1,11 +1,11 @@
 @click.command()
-@click.option('--path', '-p', required=True, help='Path to the repository.')
-@click.option('--xml_file_name', '-x', required=True, help='Name of the output XML file.')
+@click.option('--path', '-p', required=True, help='Path to the repository.', default="C:/Users/ASUS/PycharmProjects/Django/blogengine")
```
@filipp05 there must not be any references to custom absolute paths

mximp commented on main.py:
```diff
@@ -1,11 +1,11 @@
-@click.option('--path', '-p', required=True, help='Path to the repository.')
-@click.option('--xml_file_name', '-x', required=True, help='Name of the output XML file.')
+@click.option('--path', '-p', required=True, help='Path to the repository.', default="C:/Users/ASUS/PycharmProjects/Django/blogengine")
+@click.option('--xml_file_name', '-x', required=True, help='Name of the output XML file.', default="result_python_extended_stats.xml")
```
@filipp05 option names must be as simple as possible (but yet informative and unambiguous). In this case `xml` is unneeded information and  better name `--report-file`. (also options better use dashes for long names). Short name `-r`

mximp commented on models/ext_python_stats.py:
```diff
@@ -44,11 +68,451 @@ def get_metric_by_name(self, metric_name: str) -> Metric:
         pass
 
-    def get_xml_report(self) -> None:
+    def get_xml_report(self, filename: str) -> None:
```
@filipp05
a) avoid `getters` methods. Better name `report`
b) make the method return report data (dict with calculated metrics), not directly print it.
c) for the whole application workflow (init, calculate, print) create `__main__` method.

mximp commented on models/ext_python_stats.py:
```diff
@@ -44,11 +68,451 @@ def get_metric_by_name(self, metric_name: str) -> Metric:
+        with open(filename, "w", encoding="utf-8") as f:
+            f.write(pretty_xml)
+
+    def calculate_metrics_list(self) -> None:
```
@filipp05 this logic must be in the `report` method. No need to store the result within instance map: better return from the method.

mximp commented on models/ext_python_stats.py:
```diff
@@ -44,11 +68,451 @@ def get_metric_by_name(self, metric_name: str) -> Metric:
+        self.metrics_result["Number of Context Managers"] = self.__count_number_of_context_managers()
+        self.metrics_result["Number of Handled Exceptions"] = self.__count_number_of_handled_exceptions()
+
+    def __count_number_of_classes(self) -> int:
```
@filipp05 Let's make the code more object-oriented.

I suggest to create interface (ABC class) `ProjectMetrics` with single method `value() -> Dict['String':Any]`.
Individual metrics (or groups of metrics) must implement this interface. Each metric must be accompanied with tests (PyTest).

For example you can have: `FileMetrics` calculating number of files, LoC, extensions, etc.; `ClassMetrics` calculating number of classes, number of methods, fields, static members, etc.; and other groups.

Also for complex metrics (like CBO) you can have dedicated class `CBOMetric`

filipp05 commented:
with this commit I splitted all current metrics into relevant metrics-classes and simplified the ext_python_stats file. Actually, there are
— Maintainability Metrics [not all done yet]
— Code Structure Metrics
— Dependency and Coupling Metrics
— Project File Structure Metrics
— Average-based Metrics

What is more, I wrote tests for average-based metrics and reworked the init-report-print  model of our project

filipp05 commented:
with this commit I changed unittests in favor of pytest and wrote a few more tests: for code structures metrics, actually
